src/webui/webapplication.cpp

```diff
@@ -173,45 +108,151 @@ namespace
+            return false;
+        }
+
+        return (st.st_mode & S_IFMT) == S_IFREG;
+    }
+#endif
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
+    , m_isAltUIUsed(false)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    const QLatin1String tagPublic {"WEBAPI_PUBLIC"};
+    const QLatin1String tagPrivate {"WEBAPI_PRIVATE"};
+    for (int i = metaObject()->methodOffset(); i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (method.tag() == tagPublic)
+            m_publicAPIs << method.name();
+        else if (method.tag() == tagPrivate)
+            m_privateAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::sendWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
+        status(500, "Internal Server Error");
+        return;
+    }
+
+    if ((!m_isAltUIUsed) && (request().path.startsWith(PATH_PREFIX_IMAGES))) {
+        sendFile(QLatin1String(":/icons/") + request().path.mid(PATH_PREFIX_IMAGES.size()));
+        header(Http::HEADER_CACHE_CONTROL, MAX_AGE_MONTH);
+        return;
+    }
+
+    const QString path {
+        (request().path != QLatin1String("/")
+                ? request().path
+                : (sessionActive()
+                   ? QLatin1String("/index.html")
+                   : QLatin1String("/login.html")))
+    };
+
+    QString localPath {
+        m_rootFolder
+                + (sessionActive() ? PRIVATE_FOLDER : PUBLIC_FOLDER)
+                + path
+    };
+
+    QFileInfo fileInfo {localPath};
+
+    if (!fileInfo.exists() && sessionActive()) {
+        // try to send public file if there is no private one
+        localPath = m_rootFolder + PUBLIC_FOLDER + path;
+        fileInfo.setFile(localPath);
+    }
+
+    if (m_isAltUIUsed) {
+#ifdef Q_OS_UNIX
+        if (!isRegularFile(localPath.toUtf8().constData())) {
+            status(500, "Internal Server Error");
+            print(tr("Unacceptable file type."), Http::CONTENT_TYPE_TXT);
+            return;
+        }
+#endif
+
+        while (fileInfo.filePath() != m_rootFolder) {
+            if (fileInfo.isSymLink()) {
+                status(500, "Internal Server Error");
+                print(tr("Symlinks inside WWW root folder are forbidden."), Http::CONTENT_TYPE_TXT);
```

Comment:
Users won't know what is "WWW root folder", so "Symlinks inside alternative UI folder are forbidden."